net/device/IotHub.cs

```diff
@@ -75,10 +87,97 @@ public bool Cleanup(bool force = false)
+                    {
+                        // App is signalled to exit
+                        return;
+                    }
+                }
+
+                _logger.Metric("MessageBacklog", _messagesToSend.Count);
+
+                // If not connected, skip the work below this round
+                if (!_isConnected)
+                {
+                    _logger.Trace($"Waiting for connection before batching telemetry", TraceSeverity.Warning);
+                    continue;
+                }
+
+                // Make a batch of messages to send, unless we're retrying a previous batch
+                if (!_pendingMessages.Any()
+                    && _messagesToSend.Any())
+                {
+                    int batchSizeBytes = 0;
+
+                    while (_messagesToSend.TryPeek(out Message nextMessage))
+                    {
+                        int nextMessageSize = MessageHelpers.GetMessagePayloadSize(nextMessage);
+                        if (batchSizeBytes + nextMessageSize > MessageHelpers.MaxMessagePayloadBytes)
+                        {
+                            break;
+                        }
+
+                        batchSizeBytes += nextMessageSize;
+                        _messagesToSend.TryDequeue(out Message thisMessage);
+                        _pendingMessages.Add(thisMessage);
+                    }
+
+                    _logger.Metric("MessageBacklog", _messagesToSend.Count);
+                    _logger.Metric("BatchSizeBytes", batchSizeBytes);
+                }
+
+                // Send any messages prepped to send
+                if (_pendingMessages.Any())
+                {
+                    _logger.Metric("PendingMessages", _pendingMessages.Count);
+
+                    try
+                    {
+                        await _deviceClient.SendEventBatchAsync(_pendingMessages).ConfigureAwait(false);
+
+                        _totalMessagesSent += _pendingMessages.Count;
+                        _logger.Metric("TotalMessagesSent", _totalMessagesSent);
+
+                        foreach (var message in _pendingMessages)
+                        {
+                            message.Dispose();
+                        }
+                        _pendingMessages.Clear();
+                    }
+                    catch (IotHubException ex) when (ex.IsTransient)
+                    {
+                        _logger.Trace($"Caught transient exception; will retry: {ex}", TraceSeverity.Warning);
+                    }
+                    catch (Exception ex) when (ExceptionHelper.IsNetworkExceptionChain(ex))
+                    {
+                        _logger.Trace($"A network-related exception was caught; will retry: {ex}", TraceSeverity.Warning);
+                    }
+                    catch (Exception ex)
+                    {
+                        _logger.Trace($"Unknown error batching telemetry: {ex}", TraceSeverity.Critical);
```

Comment:
not related to batching logic - since we consume all exception types, some kind of a counter here will keep track of how many times we got a network exception, etc. 
Also, we might want to add an exit strategy -> if we get a critical exception more than 5 times in a row, we should exit/ log exception and then dispose the existing client and recreate?

Reply:
Part of the point of long-haul testing is to find these scenarios. In this case, I don't want to proactively code for a condition we haven't found.